### ISSUE TITLE
Feature Request: Label Grouping and Multi-Level Axis Labels

### ISSUE THREAD

**Issue by ljetibo** [matplotlib/matplotlib#6321]
Even though users can deal with axis labels on a higher level through the Formatter and Text classes, Matplotlib's handling of Labels is still very flat. At best users can change the string formatting through the Formatter class and text formatting (font, size, rotation...) through the Text class. 

Neither approach, however, offers users the ability to represent their data more accurately by using multiple label levels on the same plot in the way Excel, Root or R do. 

[Root](https://root.cern.ch/root/html534/guides/users-guide/Graphics.html#drawing-axis-independently-of-graphs-or-histograms) and each axes can be assigned a new Label instance which allows maximal flexibility for the user. 

[Excel](http://excel.tips.net/T003203_Two-Level_Axis_Labels.html) and [R](http://stackoverflow.com/questions/20571306/multi-row-x-axis-labels-in-ggplot-line-chart) handle this issue similarly by using the data-name in the legend, instead of as axis labels, and then use the group of the data-name as axis labels. In both cases the end result looks like:

![latex](https://cloud.githubusercontent.com/assets/6840364/14679329/ab30f0b2-0719-11e6-9cfa-e0a3f08e04cd.png)

No need to say that Root approach seems as an overkill and I'm more partial to Excel and R approach. A workaround by using hand-set text objects is provided in [this StackOverflow question](http://stackoverflow.com/questions/19184484/how-to-add-group-labels-for-bar-charts-in-matplotlib). The frequency of similar questions is the main motivation behind this request.

**Comment by tacaswell** [comment]
attn @story645 this roughly falls under the categorical umbrella.

**Comment by michalkahle** [comment]
I wish this worked with pandas multiindices.

**Comment by story645** [comment]
possibly a dupe of #1257

**Comment by story645** [comment]
Implement this as extension of formatters? Not just major/minor but arbitrary?

**Comment by anntzer** [comment]
I agree that some kind of nested formatters may be useful here.  Also note that in the original example the ticks are drawn "halfway" between where two ticks would normally be drawn by matplotlib (but not the labels), also something to look at...

**Comment by Khalilsqu** [comment]
https://stackoverflow.com/questions/52323446/graphing-a-multi-level-index-dataframe-with-pandas-seaborn another such question in stackoverflow. simple things can be don in excel but not easily in matplotlib.

**Comment by jklymak** [comment]
Pull requests always welcome.

FWIW, Excel was released 31 years ago, and is developed by a company with an operating budget on the order of a billion a year.  Yet, Excel still can't contour a 2-D matrix of data.

**Comment by jklymak** [comment]
Secondary axis will be a reasonable way to impliment the hierarchy of ticks.  

But as you say actually defining the hierarchy in the data is the problem.  I’d suggest a higher level library like pandas with well defined data hierarchies would be a better place to implement this.

**Comment by Jingxiang6** [comment]
I'm running into the same issue, district->county->A/B, can plot each district separately, but really wish there is an easy way to show everything in figure.

**Comment by paulbrodersen** [comment]
I just came across this issue when addressing [this question on SO](https://stackoverflow.com/questions/58854335/how-to-label-y-ticklabels-as-group-category-in-seaborn-clustermap/58915100#58915100). I think solving the full problem is hard and the attempt I made on SO is probably not the way to go in the long run. 

But, I think a lower level and small and easy to implement step towards addressing the full problem (without having to solve the API issues discussed by @ljetibo ) would be a means to annotate a range of x or y values on the axes (in analogy to the existing capabilities to annotate a single data point). I imagine the annotation object to consist of a rectangle and two lines delimiting the start and end (i.e. the 'ticks') with the following attributes:

- text 
- min, max (in data coordinates)  
- position ('left', 'bottom', 'right', 'top', 'x'=='bottom', 'y'=='left') 
- offset (in axis coordinates)
- width (in axis coordinates)
- facecolor (default None)
- (face-)alpha 
- linecolor (default black)
- linewidth (default tick width) 
- text properties (font, fontsize, alignment, etc) 

This would be fairly easy to implement, would immediately solve the base case, and could be a foundation on which to build a solution to the full problem.

**Comment by paulbrodersen** [comment]
For example, the following solves the case base for annotating a range of y-values along the left spine and demonstrates how this solution could scale to the hierarchical case.

![Figure_1](https://user-images.githubusercontent.com/8046146/69165087-e9d59300-0ae8-11ea-8478-ac9e7a3e0e01.png)

![Figure_2](https://user-images.githubusercontent.com/8046146/69165106-f0fca100-0ae8-11ea-9e26-036973caf0d8.png)

```python
#!/usr/bin/env python
"""
Annotate a range of y-values along the left spine.
"""

import matplotlib.pyplot as plt
import matplotlib.transforms as transforms

from matplotlib.patches import Rectangle
from matplotlib.lines import Line2D

def annotate_yrange(ymin, ymax,
                    label=None,
                    offset=0.1,
                    width=0.1,
                    ax=None,
                    facecolor='yellow',
                    facealpha=1,
                    linecolor='black',
                    linealpha=1,
                    linewidth=2,
):
    if ax is None:
        ax = plt.gca()

    # x-coordinates in axis coordinates,
    # y coordinates in data coordinates
    trans = transforms.blended_transform_factory(
        ax.transAxes, ax.transData)

    # a bar indicting the range of values
    rect = Rectangle((-offset, ymin),
                     width=-width,
                     height=ymax-ymin,
                     transform=trans,
                     clip_on=False,
                     color=facecolor,
                     alpha=facealpha,
    )
    ax.add_patch(rect)

    # delimiters at the start and end of the range mimicking ticks
    min_delimiter = Line2D((-(offset+width), -offset), (ymin, ymin),
                           linewidth=linewidth,
                           color=linecolor,
                           alpha=linealpha,
                           clip_on=False,
    )
    max_delimiter = Line2D((-(offset+width), -offset), (ymax, ymax),
                           linewidth=linewidth,
                           color=linecolor,
                           alpha=linealpha,
                           clip_on=False,
    )
    ax.add_artist(min_delimiter)
    ax.add_artist(max_delimiter)

    # label
    if label:
        x = -offset -width/2
        y = ymin + (ymax - ymin) / 2
        ax.text(x, y, label,
                rotation='vertical',
                transform=trans,
                horizontalalignment='center',
                verticalalignment='center',
                clip_on=False)


def demo():
    fig, ax = plt.subplots(1, 1)

    # add some extra space for the annotations
    fig.subplots_adjust(left=0.2)

    annotate_yrange(0.1, 0.9, 'test')

    plt.show()


def demo_hierarchy():
    fig, ax = plt.subplots(1, 1)
    fig.subplots_adjust(left=0.3)

    offsets = [0.1, 0.2]
    lower = [(0.1, 0.3), (0.3, 0.5), (0.5, 0.7), (0.7, 0.9)]
    upper = [(0.1, 0.5), (0.5, 0.9)]

    for ii, (level, offset) in enumerate(zip((lower, upper), offsets)):
        for jj, (ymin, ymax) in enumerate(level):
            annotate_yrange(ymin, ymax, f'test {ii}.{jj}', offset=offset)

    plt.show()


if __name__ == '__main__':

    demo()
    demo_hierarchy()

```

**Comment by jklymak** [comment]
Adding a second level of ticking is probably fine for core, and I agree would be easier if we did in Matplotlib.  Adding how to automatically decide on that second level of ticking from the structure of data seems too domain specific. 

I'm not sure what the second level of ticking API interface would look like though, nor what is to stop us at two levels (versus 3, ..., versus N)
